From 1890f67c9114d6fdf7696d6fa4eef62b1d7f74c4 Mon Sep 17 00:00:00 2001
From: Roger Monk <r-monk@ti.com>
Date: Wed, 10 Aug 2011 16:54:34 +0200
Subject: [PATCH 1/3] omap2-displays: Add support for ThreeFiveCorp S9700 Display

* Add support for s9700rtwv35tr01b display panel
   * This is used on BeagleBoardToys ULCD/-lite daugtercards

Signed-off-by: Roger Monk <r-monk@ti.com>
---
 drivers/video/omap2/displays/Kconfig               |    6 +
 drivers/video/omap2/displays/Makefile              |    1 +
 .../omap2/displays/panel-tfc-s9700rtwv35tr01b.c    |  118 ++++++++++++++++++++
 3 files changed, 125 insertions(+), 0 deletions(-)
 create mode 100644 drivers/video/omap2/displays/panel-tfc-s9700rtwv35tr01b.c

diff --git a/drivers/video/omap2/displays/Kconfig b/drivers/video/omap2/displays/Kconfig
index 609a280..f8be2a8 100644
--- a/drivers/video/omap2/displays/Kconfig
+++ b/drivers/video/omap2/displays/Kconfig
@@ -30,6 +30,12 @@ config PANEL_NEC_NL8048HL11_01B
 		This NEC NL8048HL11-01B panel is TFT LCD
 		used in the Zoom2/3/3630 sdp boards.
 
+config PANEL_TFCS9700RTWV35TR01B
+        tristate "TFC S9700RTWV35TR01B"
+        depends on OMAP2_DSS
+        help
+          LCD Panel used on BeagleboardToys 800x480 LCD Expansion Module
+
 config PANEL_TAAL
         tristate "Taal DSI Panel"
         depends on OMAP2_DSS_DSI
diff --git a/drivers/video/omap2/displays/Makefile b/drivers/video/omap2/displays/Makefile
index 0f601ab..8abdf50 100644
--- a/drivers/video/omap2/displays/Makefile
+++ b/drivers/video/omap2/displays/Makefile
@@ -5,4 +5,5 @@ obj-$(CONFIG_PANEL_NEC_NL8048HL11_01B) += panel-nec-nl8048hl11-01b.o
 
 obj-$(CONFIG_PANEL_TAAL) += panel-taal.o
 obj-$(CONFIG_PANEL_TPO_TD043MTEA1) += panel-tpo-td043mtea1.o
+obj-$(CONFIG_PANEL_TFCS9700RTWV35TR01B) += panel-tfc-s9700rtwv35tr01b.o
 obj-$(CONFIG_PANEL_ACX565AKM) += panel-acx565akm.o
diff --git a/drivers/video/omap2/displays/panel-tfc-s9700rtwv35tr01b.c b/drivers/video/omap2/displays/panel-tfc-s9700rtwv35tr01b.c
new file mode 100644
index 0000000..c2a1a19
--- /dev/null
+++ b/drivers/video/omap2/displays/panel-tfc-s9700rtwv35tr01b.c
@@ -0,0 +1,118 @@
+/*
+ * LCD panel driver for TFC S9700RTWV35TR-01B
+ *
+ * Copyright (C) 2011 Texas Instruments Inc
+ * Author: Roger Monk <r-monk@ti.com>
+ * From Original by : Vaibhav Hiremath <hvaibhav@ti.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <linux/module.h>
+#include <linux/delay.h>
+#include <linux/device.h>
+#include <linux/err.h>
+
+#include <video/omapdss.h>
+
+static struct omap_video_timings tfc_timings = {
+	.x_res		= 800,
+	.y_res		= 480,
+
+	.pixel_clock	= 30000,
+
+	.hsw		= 49,
+	.hfp		= 41,
+	.hbp		= 40,
+
+	.vsw		= 4,
+	.vfp		= 14,
+	.vbp		= 29,
+};
+
+static int tfc_panel_probe(struct omap_dss_device *dssdev)
+{
+	dssdev->panel.config = OMAP_DSS_LCD_TFT | OMAP_DSS_LCD_IVS |
+		OMAP_DSS_LCD_IHS; // | OMAP_DSS_LCD_IEO; - TODO check this - doesn't work with this enabled
+	dssdev->panel.acb = 0x0;
+	dssdev->panel.timings = tfc_timings;
+
+	return 0;
+}
+
+static void tfc_panel_remove(struct omap_dss_device *dssdev)
+{
+}
+
+static int tfc_panel_enable(struct omap_dss_device *dssdev)
+{
+	int r = 0;
+
+	/* wait couple of vsyncs until enabling the LCD */
+	msleep(50);
+
+	if (dssdev->platform_enable)
+		r = dssdev->platform_enable(dssdev);
+
+	return r;
+}
+
+static void tfc_panel_disable(struct omap_dss_device *dssdev)
+{
+	if (dssdev->platform_disable)
+		dssdev->platform_disable(dssdev);
+
+	/* wait at least 5 vsyncs after disabling the LCD */
+
+	msleep(100);
+}
+
+static int tfc_panel_suspend(struct omap_dss_device *dssdev)
+{
+	tfc_panel_disable(dssdev);
+	return 0;
+}
+
+static int tfc_panel_resume(struct omap_dss_device *dssdev)
+{
+	return tfc_panel_enable(dssdev);
+}
+
+static struct omap_dss_driver tfc_s9700_driver = {
+	.probe		= tfc_panel_probe,
+	.remove		= tfc_panel_remove,
+
+	.enable		= tfc_panel_enable,
+	.disable	= tfc_panel_disable,
+	.suspend	= tfc_panel_suspend,
+	.resume		= tfc_panel_resume,
+
+	.driver         = {
+		.name   = "tfc_s9700_panel",
+		.owner  = THIS_MODULE,
+	},
+};
+
+static int __init tfc_panel_drv_init(void)
+{
+	return omap_dss_register_driver(&tfc_s9700_driver);
+}
+
+static void __exit tfc_panel_drv_exit(void)
+{
+	omap_dss_unregister_driver(&tfc_s9700_driver);
+}
+
+module_init(tfc_panel_drv_init);
+module_exit(tfc_panel_drv_exit);
+MODULE_LICENSE("GPL");
-- 
1.6.6.1

